core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -582,6 +723,16 @@ public ShardRouting moveToUnassigned(ShardRouting shard, UnassignedInfo unassign
         return unassigned;
     }
 
+    private ShardRouting movePrimaryToUnassignedAndDemoteToReplica(ShardRouting shard, UnassignedInfo unassignedInfo) {
+        ensureMutable();
```

Comment:
`remove(shard)` does an `ensureMutable`, not sure if we want to be double-careful here or not